rs/registry/canister/src/mutations/node_management/do_replace_operator.rs

```diff
@@ -0,0 +1,829 @@
+        };
+        mutations.push(update(
+            old_node_operator_key,
+            updated_old_operator_record.encode_to_vec(),
+        ));
+
+        self.maybe_apply_mutation_internal(mutations);
+
+        println!(
+            "{}do_replace_operator: Finished executing payload: {:?}",
+            LOG_PREFIX, payload
+        );
+
+        Ok(())
+    }
+}
+
+fn find_node_operator_record_for_provider<'a>(
+    operators: &'a [NodeOperatorRecord],
+    operator_id: &'a PrincipalId,
+    provider: &'a PrincipalId,
+) -> Result<&'a NodeOperatorRecord, String> {
+    operators
+        .iter()
+        .find(|o| o.node_operator_principal_id == operator_id.0.as_slice())
+        .ok_or_else(|| {
+            format!(
+                "{}do_replace_operator: Operator {} not found for provider {}",
+                LOG_PREFIX, operator_id, provider
+            )
+        })
+}
+
+#[cfg(test)]
+mod tests {
+    use std::cell::RefCell;
+
+    use ic_config::crypto::CryptoConfig;
+    use ic_crypto_node_key_generation::generate_node_keys_once;
+    use ic_crypto_node_key_validation::ValidNodePublicKeys;
+    use ic_nns_test_utils::registry::create_subnet_threshold_signing_pubkey_and_cup_mutations;
+    use ic_protobuf::{
+        registry::{
+            crypto::v1::{AlgorithmId, PublicKey, X509PublicKeyCert},
+            dc::v1::DataCenterRecord,
+            node::v1::{ConnectionEndpoint, NodeRecord},
+            node_operator::v1::NodeOperatorRecord,
+            replica_version::v1::{BlessedReplicaVersions, ReplicaVersionRecord},
+            routing_table::v1::{routing_table::Entry, CanisterIdRange, RoutingTable},
+            subnet::v1::{SubnetListRecord, SubnetRecord, SubnetType},
+            unassigned_nodes_config::v1::UnassignedNodesConfigRecord,
+        },
+        types::v1::SubnetId,
+    };
+    use ic_registry_canister_api::ReplaceNodeOperatorPayload;
+    use ic_registry_keys::{
+        make_blessed_replica_versions_key, make_crypto_node_key, make_crypto_tls_cert_key,
+        make_data_center_record_key, make_node_operator_record_key, make_node_record_key,
+        make_replica_version_key, make_routing_table_record_key, make_subnet_list_record_key,
+        make_subnet_record_key, make_unassigned_nodes_config_record_key,
+    };
+    use ic_registry_transport::pb::v1::{registry_mutation::Type, RegistryMutation};
+    use ic_stable_structures::Storable;
+    use ic_types::{crypto::KeyPurpose, CanisterId, NodeId, PrincipalId, ReplicaVersion};
+
+    use crate::registry::Registry;
+    use prost::Message;
+
+    fn operator(n: u64) -> PrincipalId {
+        PrincipalId::new_user_test_id(n)
+    }
+
+    // To differentiate between `operator(1)` and `provider(1)`.
+    fn provider(n: u64) -> PrincipalId {
+        PrincipalId::new_user_test_id(u64::MAX - n)
+    }
+
+    // Convenience function for readability of
+    // the test.
+    fn caller(n: u64) -> PrincipalId {
+        provider(n)
```

Comment:
This seems like a piece of the puzzle - caller is assumed to be the node provider?

Reply:
Yes! It is the logical way to do it because:
* Providers have the identities on which they receive rewards already 
* They use HSM for those meaning that its way more secure

So the caller has to be the _node provider_ (with their HSM)